types/messaging.ts

```diff
@@ -1,4 +1,54 @@
-export interface CfrMessage {
-  slug: string;
-  trigger: object;
+// TODO: This needs expanding
+interface MessageTrigger {
+  /** Name of the trigger */
+  id: string;
+  /** Additional parameters of the trigger */
+  params?: Array<string|number>
+}
+
+export interface StandardMessageRecipe {
+  /** Unique identifier for the message */
+  id: string;
+
+  /** Identifier that routes the message definition to the right UI. */
+  template: string;
+
+  /** A JEXL expression representing targeting information */
+  targeting?: string;
+
+  trigger?: MessageTrigger;
+
+  /** If more than one message is valid for a, the message with the higher priority will be shown  */
+  priority?: number;
+
+  /**
+   * For messages types that are randomly chosen (e.g. snippets), determines the relative probabily
+   * of a message being shown. For example, a message with a weight of 200 will show up twice as
+   * often as other messages (since the default is 100).
+   *
+   * @default 100
+   */
+  weight?: number;
+
+  /* Definition of maximum impressions in a particular period, ogit tar groups of periods */
+  frequency?: {
```

Comment:
Is the idea here that you would pass either `lifetime` or `custom`, but not both? Or can they be used in tandem?

Reply:
They can used in tandem, yes – for example, you could have a lifetime cap of 3 (i.e. the message would never be shown again after 3 impressions) but also a daily cap of 1